core/services/eth/errors_test.go

```diff
@@ -191,3 +195,60 @@ func Test_Eth_Errors_Fatal(t *testing.T) {
 		})
 	}
 }
+
+// go-ethereum@v1.10.0/rpc/json.go
+type jsonError struct {
+	Code    int         `json:"code"`
+	Message string      `json:"message"`
+	Data    interface{} `json:"data,omitempty"`
+}
+
+func (err *jsonError) Error() string {
+	if err.Message == "" {
+		return fmt.Sprintf("json-rpc error %d", err.Code)
+	}
+	return err.Message
+}
+
+func Test_ExtractRevertReasonFromRPCError(t *testing.T) {
+	message := "important revert reason"
+	messageHex := utils.RemoveHexPrefix(hexutil.Encode([]byte(message)))
+	sigHash := "12345678"
+	var jsonErr error = &jsonError{
+		Code:    1,
+		Data:    fmt.Sprintf("0x%s%s", sigHash, messageHex),
+		Message: "something different",
+	}
+
+	t.Run("it extracts revert reasons when present", func(tt *testing.T) {
+		revertReason, err := eth.ExtractRevertReasonFromRPCError(jsonErr)
+		require.NoError(t, err)
+		require.Equal(t, message, revertReason)
+	})
+
+	t.Run("it unwraps wrapped errors", func(tt *testing.T) {
```

Comment:
I think this will also work with a multi-wrap but good to add a test to prove it

Reply:
:+1: